test: use `MiniWallet` in `test/functional/interface_zmq`

josibake opened the pull request:
While working on #24584 , `interface_zmq` started failing due to coin selection not running deterministically. The test doesn't actually need the wallet, so this PR migrates it to use MiniWallet

_Note for reviewers:_ the second commit moves large chunks of code out of an if block, so it may be helpful to review with something that ignores whitespace, e.g `git diff -w master`

josibake committed:
test: use MiniWallet in `interfaces_zmq`

make interfaces_zmg run deterministically.
this test is for the zmg notifications,
so it doesn't need the wallet compiled to run

MarcoFalke commented on test/functional/interface_zmq.py:
```diff
@@ -326,126 +324,126 @@ def test_sequence(self):
+        assert_equal(self.nodes[0].submitblock(block.serialize().hex()), None)
+        tip = self.nodes[0].getbestblockhash()
+        assert_equal(int(tip, 16), block.sha256)
+        orig_tx_2 = self.wallet.send_self_transfer(from_node=self.nodes[0])
```
I think you can just assign to `orig_txid_2` to minimize the diff (also elsewhere)

MarcoFalke replied:
I don't think this is fixed?

josibake replied:
where are you still seeing an issue? I renamed `orig_tx_2` -> `orig_txid_2` and also added `payment_txid` as a variable to avoid touching lines just to rename `payment_txid` -> `payment_tx['txid']`

MarcoFalke replied:
> where are you still seeing an issue?

In GitHub.

Usually GitHub will display "outdated" if a review comment was addressed.

You can also check the files tab and use your browsers search feature to see that the `orig_tx_2` still exists.

josibake replied:
sorry, you're right! it was fixed in the first commit, but not in the second commit. it's fixed now.

MarcoFalke commented on test/functional/interface_zmq.py:
```diff
@@ -326,126 +324,126 @@ def test_sequence(self):
+<<<<<<< HEAD
+>>>>>>> ec13426c0d (self_transfer)
+=======
+>>>>>>> ead059e477 ([move only] remove `is_wallet_compiled` checks)
```
?

josibake replied:
whoops! fixing

theStack reviewed (commented):
Nice, Concept ACK, will review this one soon!

Some quick findings:
Note that there's a typo in the PR title (s/zmg/zmq/), also I'm not sure if "refactor" is really applying here. For the second commit, adding a review suggestion for ignoring whitespace (not tested, but the very popular `--color-moved=dimmed-zebra` should work) would be probably very helpful for reviewers.

josibake committed:
[move only] remove `is_wallet_compiled` checks

josibake commented:
> Nice, Concept ACK, will review this one soon!
> 
> Some quick findings: Note that there's a typo in the PR title (s/zmg/zmq/), also I'm not sure if "refactor" is really applying here. For the second commit, adding a review suggestion for ignoring whitespace (not tested, but the very popular `--color-moved=dimmed-zebra` should work) would be probably very helpful for reviewers.

thanks for the suggestions, @theStack ! I updated the title and also added a note for reviewers. I found that `git diff -w master` worked best for me

vincenzopalazzo reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/24653/commits/bc90b8d86916d43867762a391633664676550bd8

MarcoFalke commented on test/functional/interface_zmq.py:
```diff
@@ -455,10 +452,10 @@ def test_mempool_sync(self):
         # Some transactions have been happening but we aren't consuming zmq notifications yet
         # or we lost a ZMQ message somehow and want to start over
-        txids = []
+        txs = []
```
Unrelated: It looks like this array is unused beside the last value? Could remove it and just assign the last value to a non-array name?

josibake replied:
agreed, was very confused by this array. ended up leaving it alone just to minimize my changes to just `MiniWallet`

MarcoFalke replied:
Yes, it is good to keep unrelated changes for separate commits and/or follow-ups.

MarcoFalke commented on test/functional/interface_zmq.py:
```diff
@@ -484,11 +481,12 @@ def test_mempool_sync(self):
-        self.nodes[0].bumpfee(txids[-1])
+            txs.append(self.wallet.send_self_transfer(from_node=self.nodes[0]))
+        txs[-1]['tx'].vout[0].nValue -= 1000
+        self.nodes[0].sendrawtransaction(txs[-1]['tx'].serialize().hex())
```
I recall that you asked on IRC whether it makes sense to introduce a bumpfee method for the wallet. On a second thought, I think it could make sense.

Currently it looks like `send_self_transfer` will put an utxo in the wallet which will never exist if the tx is bumped?

So a helper could drop the utxo(s?) and even add the new ones?

josibake replied:
yep, I think it would be a useful method to have. I'll open a follow-up PR for adding it as a method to `MiniWallet`

maflcko reviewed (commented):
Left some thoughts
